harvesters/src/main/java/org/fao/geonet/kernel/harvest/harvester/database/DatabaseHarvesterAligner.java

```diff
@@ -0,0 +1,408 @@
+    }
+
+    @Override
+    public HarvestResult harvest(Logger log) throws Exception {
+        this.log = log;
+        if (log.isDebugEnabled()) {
+            log.debug("Retrieving remote metadata information for : " + params.getName());
+        }
+
+        DatabaseMetadataRetriever metadataRetriever = DatabaseMetadataRetrieverFactory.getDatabaseMetadataRetriever(params.getDatabaseType(), params.getServer(), params.getPort(),
+            params.getDatabase(), params.getUsername(), params.getPassword(), log);
+
+        log.info("Start of alignment for : " + params.getName());
+        result = new HarvestResult();
+        //----------------------------------------------------------------
+        //--- retrieve all local categories and groups
+        //--- retrieve harvested uuids for given harvesting node
+        localCateg = new CategoryMapper(context);
+        localGroups = new GroupMapper(context);
+        localUuids = new UUIDMapper(context.getBean(IMetadataUtils.class), params.getUuid());
+
+        Pair<String, Map<String, Object>> filter = HarvesterUtil.parseXSLFilter(params.getXslfilter());
+        processName = filter.one();
+        processParams = filter.two();
+
+
+        metadataManager.flush();
+
+        idsForHarvestingResult = new ArrayList<>();
+
+        metadataRetriever.processMetadata(cancelMonitor, params, this);
+
+        //
+        // delete locally existing metadata from the same source if they were
+        // not in this harvesting result
+        //
+        deleteLocalMetadataNotInDatabase(idsForHarvestingResult);
+
+        return result;
+    }
+
+    public void align(String metadata) {
+        if (cancelMonitor.get()) {
+            return;
+        }
+
+        try {
+            result.totalMetadata++;
+
+            if (!StringUtils.hasLength(metadata)) {
+                log.info("Processing empty metadata xml. Skipping");
+                return;
+            }
+
+            // create JDOM element from String-XML
+            Element metadataElement = Xml.loadString(metadata, false);
+            String id = processMetadata(metadataElement);
+
+            if (StringUtils.hasLength(id)) {
+                idsForHarvestingResult.add(Integer.valueOf(id));
+            }
+
+        } catch (Exception ex) {
+            log.error("Unable to process record from database (" + this.params.getName() + ")");
+            log.error("   Record failed. Error is: " + ex.getMessage());
+            log.error(ex);
+            errors.add(new HarvestError(this.context, ex));
+        } finally {
+            result.originalMetadata++;
+        }
+    }
+
+    private void deleteLocalMetadataNotInDatabase(List<Integer> idsForHarvestingResult) throws Exception {
+        Set<Integer> idsResultHs = Sets.newHashSet(idsForHarvestingResult);
+        List<Integer> existingMetadata = context.getBean(MetadataRepository.class).findIdsBy(MetadataSpecs.hasHarvesterUuid(params.getUuid()));
```

Comment:
Could a class member be made from MetadataRepository bean ?

Reply:
Done in https://github.com/geonetwork/core-geonetwork/commit/0d74819fa131f8f61455bc462420a305864d7c0c